addon/lib/Feature.jsm

```diff
@@ -1,91 +1,577 @@
+          this._removeListenersFromWindow(subject);
+        }
+        break;
+    }
+  }
+
+  onLocationChange(browser, webProgress, request, locationURI, /* aFlags */) {
+    // Note: a null request probably means this was just a simple session restore.
+    if (locationURI && locationURI.spec && request && webProgress.isTopLevel) {
+      const topWindow = Services.wm.getMostRecentWindow("navigator:browser");
+      if (topWindow && topWindow.gBrowser) {
+        const tab = topWindow.gBrowser.getTabForBrowser(browser);
+        if (tab && tab.selected) {
+          this._checkDocument(locationURI.spec);
+        }
+      }
+    }
   }
 
+  handleEvent(event) {
+    const eventPrefix = this.shownPanelType ? this.shownPanelType + "-" : "";
+    switch (event.type) {
+      case "command": {
+        // 'Okay, got it' button was clicked.
+        Services.prefs.setBoolPref(PREF_NUDGES_DISMISSED_WITHOK, true);
+        let panel = event.target.parentNode;
+        while (panel.parentNode && panel.localName != "panel") {
+          panel = panel.parentNode;
+        }
+        // Hide the panel when the button is clicked.
+        if (panel && panel.hidePopup) {
+          panel.hidePopup();
+          this.telemetry({event: eventPrefix + "hidden-buttonclick"});
+        }
+        break;
+      }
+      case "popuphidden": {
+        // Check if the panel was hidden by clicking the URLBar.
+        const window = event.target.ownerGlobal;
+        const focusMethod = Services.focus.getLastFocusMethod(window);
+        if (window.gURLBar.focused && focusMethod && !!(focusMethod & Services.focus.FLAG_BYMOUSE)) {
+          Services.prefs.setBoolPref(PREF_NUDGES_DISMISSED_CLICKAB, true);
+          this.telemetry({event: eventPrefix + "hidden-awesomebarclick"});
+        } else {
+          this.telemetry({event: eventPrefix + "hidden"});
+        }
+        this.shownPanelType = null;
+        break;
+      }
+      case "TabSelect": {
+        if (event.target.linkedBrowser &&
+            event.target.linkedBrowser.currentURI) {
+          this._checkDocument(event.target.linkedBrowser.currentURI.spec);
+        }
+        break;
+      }
+      default:
+        Cu.reportError(`ShieldSearchNudges: Unknown event: ${event.type}`);
+        break;
+    }
+  }
+
+  /**
+   * On startup, we need to check that the default engine isn't being displayed,
+   * and if it isn't then we display the doorhanger if one of the pages we
+   * are looking for is displayed.
+   */
+  _checkDisplayOnFirstStartup() {
+    const winEnum = Services.ww.getWindowEnumerator();
+    while (winEnum.hasMoreElements()) {
+      const win = winEnum.getNext();
+      if (win.document.documentURI == "chrome://global/content/commonDialog.xul" ||
+          win.document.documentURI == "chrome://global/content/selectDialog.xul") {
+        // There is some sort of modal dialog displaying, probably the default
+        // browser one, so just get outta here.
+        return;
```

Comment:
I'm possibly a little nervous that we'd get into a state where we would never show it. However, as this is only on startup, then I think that it will be good enough - we might not prompt straight away, but we'll be able to do so later if the user revisits. (Hence I'll change it)